lxd/storage/drivers/driver_powerflex_utils.go

```diff
@@ -0,0 +1,963 @@
+package drivers
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/lxd/lxd/locking"
+	"github.com/canonical/lxd/lxd/resources"
+	"github.com/canonical/lxd/lxd/revert"
+	"github.com/canonical/lxd/lxd/util"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/api"
+)
+
+// powerFlexBlockVolSuffix suffix used for block content type volumes.
+const powerFlexBlockVolSuffix = ".b"
+
+// powerFlexISOVolSuffix suffix used for iso content type volumes.
+const powerFlexISOVolSuffix = ".i"
+
+// powerFlexCodeVolumeNotFound is returned by the API in case a storage volume does not exist.
+const powerFlexCodeVolumeNotFound = 79
+const powerFlexCodeNameTooLong = 226
+
+type powerFlexVolumeType string
+type powerFlexSnapshotMode string
+
+const powerFlexVolumeThin powerFlexVolumeType = "ThinProvisioned"
+const powerFlexVolumeThick powerFlexVolumeType = "ThickProvisioned"
+
+const powerFlexSnapshotRW powerFlexSnapshotMode = "ReadWrite"
+
+// powerFlexVolTypePrefixes maps volume type to storage volume name prefix.
+// Use smallest possible prefixes since PowerFlex volume names are limited to 31 characters.
+var powerFlexVolTypePrefixes = map[VolumeType]string{
+	VolumeTypeContainer: "c",
+	VolumeTypeVM:        "v",
+	VolumeTypeImage:     "i",
+	VolumeTypeCustom:    "u",
+}
+
+// powerFlexError contains arbitrary error responses from PowerFlex.
+type powerFlexError map[string]any
+
+// Error tries to return all kinds of errors from the PowerFlex API in a nicely formatted way.
+func (p *powerFlexError) Error() string {
+	var errorStrings []string
+	for k, v := range *p {
+		errorStrings = append(errorStrings, fmt.Sprintf("%s: %v", k, v))
+	}
+
+	return strings.Join(errorStrings, ", ")
+}
+
+// ErrorCode extracts the errorCode value from a PowerFlex response.
+func (p *powerFlexError) ErrorCode() float64 {
```

Comment:
Why is this a float64?

Reply:
Hope I remember correctly :) The root cause is `type powerFlexError map[string]any`. In case a response from PowerFlex contains the `errorCode` field, json unmarshal will parse the integer code as `float64`. 

I was first using dedicated return type structs but found that the API responses can be quite different. I then went with `map[string]any` which just serves the few fields we need to validate the response.

Reply:
Makes sense, could you add that as a comment to where its defined to record that reason for posterity?